modules/core/src/com/haulmont/cuba/core/app/EntitySnapshotManager.java

```diff
@@ -336,4 +312,92 @@ protected String processSnapshotXml(String snapshotXml, Map<Class, Class> classM
         replaceInXmlTree(document.getRootElement(), classMapping);
         return document.asXML();
     }
+
+    @Override
+    public EntitySnapshot getLastEntitySnapshot(Entity entity) {
+        MetaClass metaClass = extendedEntities.getOriginalOrThisMetaClass(entity.getMetaClass());
+        View view = metadata.getViewRepository().getView(EntitySnapshot.class, "entitySnapshot.browse");
+        LoadContext<EntitySnapshot> lx = LoadContext.create(EntitySnapshot.class).setQuery(LoadContext
+                .createQuery(format("select e from sys$EntitySnapshot e where e.entityMetaClass = :metaClass and"
+                                + " e.entity.%s = :entityId order by e.createTs desc",
+                        referenceToEntitySupport.getReferenceIdPropertyName(metaClass)))
+                .setParameter("metaClass", metaClass.getName())
+                .setParameter("entityId", entity.getId())
+                .setMaxResults(1))
+                .setView(view);
+        return dataManager.load(lx);
+    }
+
+    @Override
+    public EntitySnapshot getLastEntitySnapshot(MetaClass metaClass, Object id) {
+        MetaClass originalMetaClass = extendedEntities.getOriginalOrThisMetaClass(metaClass);
+        View view = metadata.getViewRepository().getView(EntitySnapshot.class, "entitySnapshot.browse");
+
+        Transaction tx = persistence.createTransaction();
+        try {
+            EntityManager em = persistence.getEntityManager();
+            //noinspection unchecked
+            Entity entity = (Entity) em.find(originalMetaClass.getJavaClass(), id);
```

Comment:
Why do we need to find it?

Reply:
Entity for which we get the last snapshot shouldn't have a composite key. And in the query for snapshots, we have to use a variable name of entity's id.